backend/zoho/zoho.go

```diff
@@ -0,0 +1,1240 @@
+// Package zoho provides an interface to the Zoho Workdrive
+// storage system.
+package zoho
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/rclone/rclone/lib/encoder"
+	"github.com/rclone/rclone/lib/pacer"
+	"github.com/rclone/rclone/lib/random"
+	"github.com/rclone/rclone/lib/readers"
+
+	"github.com/rclone/rclone/backend/zoho/api"
+	"github.com/rclone/rclone/fs"
+	"github.com/rclone/rclone/fs/config"
+	"github.com/rclone/rclone/fs/config/configmap"
+	"github.com/rclone/rclone/fs/config/configstruct"
+	"github.com/rclone/rclone/fs/config/obscure"
+	"github.com/rclone/rclone/fs/fserrors"
+	"github.com/rclone/rclone/fs/hash"
+	"github.com/rclone/rclone/lib/dircache"
+	"github.com/rclone/rclone/lib/oauthutil"
+	"github.com/rclone/rclone/lib/rest"
+	"golang.org/x/oauth2"
+)
+
+const (
+	rcloneClientID              = "1000.96D8SL83Y7ZVH86WGNDHR8IGFWR12I"
+	rcloneEncryptedClientSecret = "QszU1YQ3HV2nYT32QljIhMHy3x1bpPmQ92lyI-_xkY7T6ZsWsAkeBK_Ta9JwHg05t2CkZasvAYMN8w"
+	minSleep                    = 10 * time.Millisecond
+	maxSleep                    = 2 * time.Second
+	decayConstant               = 2 // bigger for slower decay, exponential
+	rootURL                     = "https://workdrive.zoho.com/api/v1"
+	accountsURL                 = "https://accounts.zoho.com"
+	configRootID                = "root_folder_id"
+)
+
+// Globals
+var (
+	// Description of how to auth for this app
+	oauthConfig = &oauth2.Config{
+		Scopes: []string{
+			"aaaserver.profile.read",
+			"WorkDrive.team.READ",
+			"WorkDrive.workspace.READ",
+			"WorkDrive.files.ALL",
+		},
+		Endpoint: oauth2.Endpoint{
+			AuthURL:   "https://accounts.zoho.com/oauth/v2/auth",
+			TokenURL:  "https://accounts.zoho.com/oauth/v2/token",
+			AuthStyle: oauth2.AuthStyleInParams,
+		},
+		ClientID:     rcloneClientID,
+		ClientSecret: obscure.MustReveal(rcloneEncryptedClientSecret),
+		RedirectURL:  oauthutil.RedirectPublicSecureURL,
+	}
+)
+
+// Register with Fs
+func init() {
+	fs.Register(&fs.RegInfo{
+		Name:        "zoho",
+		Description: "Zoho",
+		NewFs:       NewFs,
+		Config: func(name string, m configmap.Mapper) {
+			opt := oauthutil.Options{
+				// No refresh token unless ApprovalForce is set
+				OAuth2Opts: []oauth2.AuthCodeOption{oauth2.ApprovalForce},
+			}
+			if err := oauthutil.Config("zoho", name, m, oauthConfig, &opt); err != nil {
+				log.Fatalf("Failed to configure token: %v", err)
+			}
+			// We need to rewrite the token type to because Zoho wants
```

Comment:
Ah, the wonderful thing about standards! Nearly every backend that uses oauth has some oauth related bodge.